pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/CloseResourceRule.java

```diff
@@ -175,7 +175,7 @@ private void checkForResources(ASTMethodOrConstructorDeclaration methodOrConstru
         Map<ASTVariableDeclarator, TypeNode> resVars = new HashMap<>();
         for (ASTVariableDeclarator var : vars) {
             TypeNode varType = getTypeOfVariable(var);
-            if (isResourceTypeOrSubtype(varType)) {
+            if (varType != null && isResourceTypeOrSubtype(varType)) {
```

Comment:
So we don't get a NPE anymore, but we also don't consider the var as a violation? I wonder, for example, what is the behavior of the rule if you write
```java
import java.io.*;
public class Foo {
    public int bar() throws IOException {
        var inputStream = getInputStreamFromSomewhere();
        char c = reader.read();
        return c;
    }

    InputStream getInputStreamFromSomewhere() { return null; }
}
```

Normally, it should be exactly the same behavior as if `var` was explicitly written, right? In that case, I'm not sure the rule accounts for that, given we test the ASTType node, not VariableDeclaratorId::getType (which is the type of the var). (Note: since we do that, we are probably failing if the var is declared like `InputStream streamArray[]`)

I think it would make more sense to test VariableDeclaratorId::getType, though I don't know how that interacts with existing tests

Reply:
I was passing by and wanted to ask how is it possible to get var Type in PMD? In that case `var` variable is just ignored, right?

Reply:
If the variable type is inferred, then there is no type node, but the type of the VariableDeclaratorId node may be set anyway, if we can resolve the type of the initializer. In the example above, if `Foo` is on the classpath, then `varId.getType() == InputStream.class`

The type of the VariableDeclaratorId is always the "real" type of the variable, provided it's on the classpath and we can resolve it. Manipulating the VariableDeclaratorId is always preferable to testing the type node manually, for reasons explained in [the javadoc](https://docs.pmd-code.org/apidocs/pmd-java/6.27.0/net/sourceforge/pmd/lang/java/ast/ASTVariableDeclaratorId.html#getType())

Reply:
Thanks for your explanation!

Reply:
@oowekyala I've added a simple test case for var and it seems to work *as long as the initializer expression is not a method call*. But I think, the lack of resolving method call result types is part of PMD 6. Maybe you fixed this already on the 7 branch?

Anyway: The goal of this PR was to fix the NPE, it was never a goal to improve the rule.
Feel free to create a new issue for tracking the false negative with the above sample.

Reply:
OK in that case I believe this pr is good to merge. I'll open an issue as you said